8367319: Add os interfaces to get machine and container values separately

caspernorrbin opened the pull request:
Hi everyone,

The current `os::` layer on Linux hides whether the JVM is running inside a container or not. When running inside a container, we replace machine values with container values where applicable, without telling the user of these methods. For most use cases, this is fine, users only care about the returned value. But for other use cases, where the value originated is important. Two examples:

- A user might need the physical cpu count of the machine, but `os::active_processor_count()` only returns the limited container value, which also represents something slightly different.
- A user might want the container memory limit and the physical RAM size, but `os::physical_memory()` only gives one number.

To solve this, every function that mixed container/machine values now has to explicit variants, prefixed with `machine_` and `container_`. These use the bool return + out-parameter interface, with the container functions only working on Linux. The original methods remain and continue to return the same mixed values.

In addition, container-specific accessors for the memory soft limit and the memory throttle limit have been added, as these values matter when running in a containerized environment.

`OSContainer::active_processor_count()` has also been changed to return `double` instead of `int`. The previous implementation rounded the quota/period ratio up to produce an integer for `os::active_processor_count()`. Now, when the value is requested directly from the new container API it makes more sense to preserve this fraction rather than rounding it up. We can thus keep the exact value for those that want it, then round it up to keep the same behavior in `os::active_processor_count()`.

Testing:
- Oracle tiers 1-5
- Container tests on cgroup v1 and v2 hosts.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8367319](https://bugs.openjdk.org/browse/JDK-8367319): Add os interfaces to get machine and container values separately (**Enhancement** - P4)


### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**) Review applies to [3eda822e](https://git.openjdk.org/jdk/pull/27646/files/3eda822e28465c29dc848ee5a314992d753cc222)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/27646/head:pull/27646` \
`$ git checkout pull/27646`

Update a local copy of the PR: \
`$ git checkout pull/27646` \
`$ git pull https://git.openjdk.org/jdk.git pull/27646/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 27646`

View PR using the GUI difftool: \
`$ git pr show -t 27646`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/27646.diff">https://git.openjdk.org/jdk/pull/27646.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/27646#issuecomment-3371511667)
</details>

albertnetymk commented:
> For most use cases, this is fine, users only care about the returned value. But for other use cases, where the value originated is important.

Then, is it possible to keep the original API for "most use cases"? For others, they can query `is_containerized` and new `machine_` APIs.

The current proposed API kind of forces all users to distinguish btw inside a container or not, even though most use cases don't care.

caspernorrbin commented:
> Then, is it possible to keep the original API for "most use cases"? For others, they can query `is_containerized` and new `machine_` APIs.
> 
> The current proposed API kind of forces all users to distinguish btw inside a container or not, even though most use cases don't care.

The original API is untouched, so "most use cases" can still use that without ever worrying about containers. The `machine_` functions is simply an add-on to that API. Both the examples above remain unchanged, `os:: active_processor_count()` still returns both machine/container values, and `os::physical_memory()` still returns the appropriate memory limit.

dholmes-ora commented:
> A user might need the physical cpu count of the machine, but os::active_processor_count() only returns the limited container value, which also represents something slightly different.

That is a mis-characterization of the API. `active_processor_count()` tells you how many logical processors are available to the JVM process. That can be very different to the "physical" (**) number of processors due to partitioning at various levels (e.g. virtualization, containerization), as well as direct restrictions through API's like `taskset`.

(**) "physical" actually has no meaning these days. There is some value you can obtain through the operating system that provides the maximum number of processors that the operating system can see (and thus make available to the JVM).

dholmes-ora commented:
> By contrast, the current container-reported value treats cpu quota and logical processors as the same thing, even though quota only restricts cpu time, not the number of cores we can run on. With a quota of 1, we might still execute on two cores for 50% of the time each, but os::active_processor_count() still only reports "1".

Right - the way the "container folk" dealt with quotas etc never made any sense to me at all. I tried arguing the point but to no avail. So basically what you are looking for here is a way to get around the "broken" definition of available-processors when quotas are enforced.

jerboaa commented:
> > This will conflict mightily with the refactoring that I'm working on for https://bugs.openjdk.org/browse/JDK-8365606
> 
> Understood, thanks for flagging that. The amount of change in the container layer isn't that big except for `CgroupUtil::processor_count()`, which wasn't of a java type to begin with. Is there something I could change to lessen the conflict at all?

Separating out the `processor_count()` change would be good (irrespective of the conflict). But other than that, I'd think there isn't really a good way. Perhaps to wait for JDK-8365606 to land as those are way more changes to deal with rather than what is done here. I.e. merge should be simpler. I should have something ready by end of the week, fwiw.

dholmes-ora commented:
> What I'm after is a clean way to get those numbers separately. Pairing machine_ and container_ gives access to both pieces of data, offering just the machine side would leave us without a direct path to the container value.

I don't like the naming "machine_" here - yes naming is hard and subjective - these "machine" API's are reporting whatever the operating system is exposing to the process.

caspernorrbin commented:
> I don't like the naming "machine_" here - yes naming is hard and subjective - these "machine" API's are reporting whatever the operating system is exposing to the process.

I'm open to changing the naming to make it more clear which value is accessed. Could maybe prefix them with `os_` instead. That could however look a bit weird with a double `os::os_`.

dholmes-ora commented:
>  In other words - are you against the idea of having an implicit API that gives you either the container or the "machine"?

@fisk I am "against" trying to shoe-horn a poorly-defined  legacy API into a dichotomy of "container value" versus "machine value". The concepts are not at all clean or well-defined for both variants.

> CPU quotas when running in a container.

Then you need an explicit API for that. No contortion of "available processors" is going to tell you what your quota is.

> The "machine" numbers when running in a container (which get overridden by container numbers).

Sounds simple but is it well-defined? Can you even ask the question (again I come back to available processors where sched_getaffinity will already account for the presence of the container if tasksets are used - what answer would you want for the "machine"?). 

Aside: even if you can ask the question what use are these values if you are running within the container? Is there some means to bypass the container constraints?

fisk commented:
> Looking again, and ignoring some earlier comments by others that might have mislead me, what is being proposed is much closer to that than what I initially thought. What I wanted to avoid was having 3 variants of a given method when the original os:: form of that method isn't even well-defined today. But I am okay with a set of "machine/system" methods to ask queries that make sense for the "machine"; and a set of container ones that make sense for containers. (And the implementation of the os:: method can delegate to whichever provides the "right" answer.)

Yeah that sounds good.

> But to clarify my issue with the container notion of available processors ... you state:
> 
> > It exposes what number of processors, on average, the container is limited to run on.
> 
> Yes - but if I am asking how many processors are available I am doing so because I want to know how many threads it makes sense creating, so I know how many threads could potentially be executing in parallel at the same time. Using my example above that could be 50% of the physical cores or 100% depending on how the container decides to implement quotas. That fits in, I think, with the example Casper gave above:
> 
> > In latency-sensitive workloads we might burst onto all 16 cores for a short interval and still stay within the 2-cpu quota. At the moment, the JVM has no way to know that those extra 14 cores even exist, so it cannot make that optimization.
> 
> Indeed - but if you create 16 threads to take advantage of that and the container only actually gives you 2 cores for 100% of the time, then you have totally messed up. This notion of "average processors" does not help you size accordingly, nor does knowing the number of "machine" processors, if you don't know how the container operates. That is one assumption about the container implementation I was referring to. 

Right - the number of threads you want to use is different if you have a concurrent system, compared to a parallel system, when the container implementation shares cores across containers. With a long running parallel workload, you want to size your threads according to the container average, as you say, because you won't get more CPU than that anyway and you will only run into trouble blowing the limits. Conversely, with a concurrent workload such as a server, where the unit of work takes, say, 1 ms to execute, you are more likely to want to size the threads of your thread pool according to the machine limits. Any time a request comes in to a server and there are idle CPUs on the machine, you are better off making sure the cores are working to reduce latency, compared to just waiting around. Then other means (such as load balancing, provisioning more containers, etc) need to be used to control that the average CPU utilization does not get close to the container limits (or indeed machine limits), because that will send latencies into a roller coaster.

In other words, with a parallel workload you want to size the threads according to the container limits. With a concurrent workload you want to size according to the machine limits, but monitor that average CPU utilization both on the container and machine levels, do not get too close to their corresponding limits. And that is why we want to be able to be able to access both of them.

I guess your concern is about an alternative container implementation where the limits are strict and you get a single exclusive core that none of the other containers is allowed to use, not one core on average. Then you would not need to worry about the machine layer, due to the lack of sharing of hardware resources across containers. I think there is a philosophical question there: is that alternative container implementation really implementing a container, or a virtual machine? For all practical purposes, it seems to me like that would be running in a virtual machine with dedicated hardware that we happen to call a container. I suspect that should a container environment like that pop up, we might want to model it like a virtual machine and not like a container.

If we did model it like a container, then yes, I wouldn't want to look around at the completely orthogonal "machine" layer any longer, as I don't share hardware with any of the other "containers", and they don't share any of my hardware either. It would be a rather different situation.

> The other is that the container never defines cpusets, otherwise you have no way that I am aware of to actually determine there are 16 cores on the machine**. Maybe these are safe assumptions to make, but it would be nice if the underlying container subsystem could give you those answers directly based on how it does things (then you wouldn't need to query the "machine".)
> 
> ** If you assume no hot-swapping of CPUs then `sysconf(_SC_NPROCESSORS_ONLN)` could give you that answer.

I agree. :)

dholmes-ora commented:
> is that alternative container implementation really implementing a container, or a virtual machine?

You would have to define what you mean by "container" and "virtual machine".  Isolating CPU usage ala tasksets seems a natural partitioning approach for a "container" environment to me.

dholmes-ora commented:
>  I'd vote for crossing that bridge when we get there 

I guess. Though I still don't see why it would not be reasonable to e.g. run a container in a taskset.

caspernorrbin commented:
I want to thank everyone for the great input and thoughtful discussion. I really appreciate the depth of feedback.

I've moved ahead with inner `os::Machine` and `os::Container` classes. This keeps the `os::` methods unchanged for anyone who doesn't care about the source, but makes it much clearer how to get either the OS/system view or the container's limits when you need that distinction.

Of course, there's no perfect split here. Plenty of ambiguity remains when it comes to virtualization, affinity masks, and what we call "the machine". I have added comments and documentation try to spell out exactly what these classes represent, what each method reports, what's affected by containers and what isn't, and the best context for using each one.

For the moment, this only targets cgroup-based container environments, but I'm open to revisiting things if other environments gain traction in the future.

caspernorrbin commented:
After some offline discussion with @fisk , I've pushed a couple minor changes:
- First, when there is no memory limit, internal functions currently set the limit to a predefined `value_unlimited`. However, for callers of the public API, returning this value is less helpful, they only need to know whether a limit exists. So now, if the memory limit is unlimited, we simply return `false` to indicate there is no enforced limit.
- Second, I've removed the asserts for `is_containerized()`. There's nothing actually preventing someone from checking cgroup values directly, so we shouldn't block this with an assert. Also, since being "containerized" is partly defined as having a resource limit set, it seems redundant to require checking for a limit before accessing its value.

 
@dholmes-ora, could you please take another look at these changes? I hope I've addressed some of the earlier pain points and that we can move closer to getting this finalized.

jerboaa commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -216,22 +216,27 @@ bool os::is_containerized() {
+  if (OSContainer::memory_throttle_limit_in_bytes(value) && value != value_unlimited) {
+    return true;
+  }
+  return false;
```
Same here with the changed `value` when `value_unlimited` is being read, and function returning `false`.

jerboaa commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -216,22 +216,27 @@ bool os::is_containerized() {
+  if (OSContainer::memory_limit_in_bytes(value) && value != value_unlimited) {
+    return true;
+  }
+  return false;
```
This started with:
https://bugs.openjdk.org/browse/JDK-8357086

They should have the same meaning throughout. Returning `true` means the passed in reference has been set to a value. In the container case it might be `value_unlimited` (for example an interface file has value `max`). So if the desire is to return `false` for unlimited, then this needs to be handled on top.

caspernorrbin replied:
From my perspective, returning `true` from one of these functions means that the out-parameter is set to a valid, meaningful value and can be used safely by the caller. However, if the function returns `false`, there is no guarantee that the out-parameter is untouched, so relying on its previous value is unsafe. Your example is to me is a bad pattern and is brittle at best.

There is a distinction between `OSContainer::memory_limit_in_bytes` (which always fetches and returns the cgroup, including `value_unlimited` for no limit) and the higher-level `os::Container::memory_limit`, which is about determining whether there is an effective limit at all, and if so, what it is. It doesn't make sense to return `true` if there is no limit, since there wouldn't be a meaningful value to provide. `value_unlimited` is just an internal constant and not helpful for callers.

That said, I can update so we don't pass through the out-parameter so that we don't modify it when returning `false`, to be somewhat more consistent. But I still believe we should return `false` when no limit is set.

caspernorrbin replied:
> I think that should be guaranteed. This is our API we can make it as strict as we desire. If I have preset a sentinel value and the op fails then I should be able to rely on my sentinel remaining intact.

From the feedback I've changed so returning `false` now doesn't change the value, so we keep that assumption.

> But now you are overloading the use of the return value. It was intended to reflect success or failure, not provide semantic information. The higher-level function can choose to return true when there is a limit and false when there is not, but it can't also mean "or an error occurred".

I'd argue that in this context, we fail (and should return `false`) when we can't find a limit. When a caller asks the generic `Container` layer for a limit and none is found, "no limit" isn't a meaningful limit to report. Not finding a limit is the failure here, not being unable to read a cgroup file.

dholmes-ora replied:
Based on that you would not want errors to be reported at all. I was expecting at least logging on error or perhaps some kind of "abort" if the error signifies a fatal condition for what you are trying to do. Maybe errors are so unlikely that we really don't care? But I would not assume an error means there is no limit and proceed on that basis - YMMV.

But I will step aside from the discussion.

jerboaa commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -342,6 +342,54 @@ class os: AllStatic {
+  // If there is no limit, or the limit could not be determined, it instead returns false.
+  // Methods return true and set the out-parameter if a limit is found,
+  // or false if no limit exists or it cannot be determined.
+  class Container : AllStatic {
```
Good.
